rfc/2020/arm-support/aliariff.md

```diff
@@ -0,0 +1,267 @@
+      gcr.io/linkerd-io/proxy-init-arm64:... \
+      gcr.io/linkerd-io/proxy-init-arm:...
+    ```
+  - Use the `GOARCH` argument to perform cross compile in `Dockerfile`
+    ```Dockerfile
+    ARG GOARCH
+    RUN CGO_ENABLED=0 GOOS=linux GOARCH=$GOARCH go build ...
+    ```
+  - Change the runtime image in `Dockerfile`
+    ```Dockerfile
+    FROM --platform=$GOARCH debian:stretch-20190812-slim
+    ```
+  - Add Github actions
+  - Call the build inside github workflow
+     ```yml
+     make build # this will build the amd64 image
+     DOCKER_BUILD_ARCH=arm64 make build # building arm64 image
+     DOCKER_BUILD_ARCH=arm make build # building arm image
+     ```
+  - Push image & push multi arch manifest
+    ```yml
+    make push-all
+    ```
+
+- `linkerd2-proxy`
+  - Add new step in `.github/workflows/release.yml`
+  - Install requirement to perform cross compilation in Rust
+    ```yml
+    sudo apt-get install -y gcc-arm-linux-gnueabihf gcc-aarch64-linux-gnu
+    ```
+  - Set the appropriate environment
+    ```yml
+    env:
+      CARGO_TARGET_ARCH: armv7-unknown-linux-gnueabihf # for arm
+      CARGO_RELEASE: "1"
+    run: make package
+
+    or
+
+    env:
+      CARGO_TARGET_ARCH: aarch64-unknown-linux-gnu # for arm64
+      CARGO_RELEASE: "1"
+    run: make package
+    ```
+  - Change the Makefile to check for `CARGO_TARGET_ARCH` to perform cross compile
+    ```makefile
+    .
+    .
+    CARGO_BUILD = $(CARGO) build --frozen $(RELEASE) --target=$(CARGO_TARGET_ARCH)
+    .
+    .
+    ```
+  - Publish the artifact together with other architecture
+
+- `linkerd2`
+
+  There are 7 images to build: `controller`, `cli`, `cni-plugin`, `web`, `proxy`, `debug`, `grafana`
+
+  - Export `DOCKER_BUILD_ARCH` in `bin/_docker.sh` & append the architecture in the repo name
+    ```sh
+    export DOCKER_BUILD_ARCH="${DOCKER_BUILD_ARCH:-}"
+    .
+    .
+    .
+    docker_repo() {
+      .
+      .
+      .
+      if [ -n "${DOCKER_BUILD_ARCH:-}" ]; then
+        name="$name-$DOCKER_BUILD_ARCH"
+      fi
+      .
+      .
+      .
+    }
+    ```
+  - Pass the architecture argument to docker build in every `docker-build-*` file
+    ```sh
+    docker_build .... --build-arg "GOARCH=${DOCKER_BUILD_ARCH:-amd64}"
+    ```
+  - Use the `GOARCH` argument to perform cross compile in all `Dockerfile`
+    ```Dockerfile
+    ARG GOARCH
+    RUN CGO_ENABLED=0 GOOS=linux GOARCH=$GOARCH go build ...
+    ```
+  - Change the runtime image in `Dockerfile`
+    ```Dockerfile
+    FROM --platform=$GOARCH ...
+    ```
+  - Adjust `bin/fetch-proxy` file to retrieve the correct architecture version
+  - Add `bin/docker-manifest` file to create and push multi-arch manifest
+    ```sh
+    for img in cli-bin cni-plugin controller debug grafana proxy web ; do
+      docker manifest create gcr.io/linkerd-io/$img:$tag \
+        gcr.io/linkerd-io/$img-amd64:$tag \
+        gcr.io/linkerd-io/$img-arm64:$tag \
+        gcr.io/linkerd-io/$img-arm:$tag
```

Comment:
I assume we need a `docker manifest push` command after this.

Reply:
You are right missing the push part. Will add it.